File: android/src/main/kotlin/com/example/multi_ads/pangle_global/NativeAdManager.kt
package com.example.multi_ads.pangle_global

import android.app.Activity
import android.content.Context
import android.util.Log
import io.flutter.plugin.common.EventChannel
import com.bytedance.sdk.openadsdk.api.nativeAd.PAGNativeAd
import com.bytedance.sdk.openadsdk.api.nativeAd.PAGNativeAdLoadListener
import com.bytedance.sdk.openadsdk.api.nativeAd.PAGNativeRequest

class NativeAdManager(
    private val context: Context,
    private val activity: Activity?,
    private val listenerId: String,
    private val slotId: String,
    private val style: Map<String, Any>?,
    private var eventSink: EventChannel.EventSink?
) {
    companion object {
        const val TAG = "PangleNativeAdManager"
    }

    var nativeAd: PAGNativeAd? = null
        private set

    fun updateEventSink(sink: EventChannel.EventSink?) {
        this.eventSink = sink
    }

    fun loadAd() {
        Log.d(TAG, "Loading native ad with listenerId: $listenerId, slotId: $slotId")

        val request = PAGNativeRequest()

        PAGNativeAd.loadAd(slotId, request, object : PAGNativeAdLoadListener {
            override fun onError(code: Int, message: String?) {
                Log.e(TAG, "Native ad load failed ($listenerId): $code - $message")
                activity?.runOnUiThread {
                    eventSink?.success(mapOf(
                        "listenerId" to listenerId,
                        "event" to "onAdLoadFailed",
                        "error" to mapOf(
                            "code" to code,
                            "message" to (message ?: "Unknown error")
                        )
                    ))
                }
            }

            override fun onAdLoaded(ad: PAGNativeAd?) {
                Log.d(TAG, "Native ad loaded successfully ($listenerId)")
                nativeAd = ad

                activity?.runOnUiThread {
                    eventSink?.success(mapOf(
                        "listenerId" to listenerId,
                        "event" to "onAdLoaded"
                    ))
                }
            }
        })
    }

    fun getListenerId(): String = listenerId

    fun dispose() {
        Log.d(TAG, "Disposing ad ($listenerId)")
        nativeAd = null
    }
}

File: ios/Classes/pangle_global/NativeAdManager.swift
import Foundation
import Flutter
import PAGAdSDK

class NativeAdManager: NSObject {
    private var nativeAd: PAGLNativeAd?
    private var eventSink: FlutterEventSink?
    private let listenerId: String
    private let slotId: String
    
    init(listenerId: String, slotId: String, eventSink: FlutterEventSink?) {
        self.listenerId = listenerId
        self.slotId = slotId
        self.eventSink = eventSink
        super.init()
    }
    
    func updateEventSink(_ sink: FlutterEventSink?) {
        self.eventSink = sink
    }
    
    func loadAd() {
        print("[PangleNativeAd] Loading ad with listenerId: \(listenerId), slotId: \(slotId)")
        let request = PAGNativeRequest()
        
        PAGLNativeAd.load(withSlotID: slotId, request: request) { [weak self] ad, error in
            guard let self = self else { return }
            
            if let error = error {
                let errorCode = (error as NSError).code
                print("[PangleNativeAd] Ad load failed (\(self.listenerId)): \(errorCode) - \(error.localizedDescription)")
                DispatchQueue.main.async {
                    self.eventSink?([
                        "listenerId": self.listenerId,
                        "event": "onAdLoadFailed",
                        "error": [
                            "code": errorCode,
                            "message": error.localizedDescription
                        ]
                    ])
                }
                return
            }
            
            if let ad = ad {
                self.nativeAd = ad
                ad.delegate = self
                print("[PangleNativeAd] Ad loaded successfully (\(self.listenerId))")
                
                DispatchQueue.main.async {
                    self.eventSink?([
                        "listenerId": self.listenerId,
                        "event": "onAdLoaded"
                    ])
                }
            }
        }
    }
    
    func getNativeAd() -> PAGLNativeAd? {
        return nativeAd
    }
    
    func dispose() {
        print("[PangleNativeAd] Disposing ad (\(listenerId))")
        nativeAd = nil
    }
}

// MARK: - PAGLNativeAdDelegate
extension NativeAdManager: PAGLNativeAdDelegate {
    func adDidShow(_ ad: PAGAdProtocol) {
        print("[PangleNativeAd] Ad showed (\(listenerId))")
        DispatchQueue.main.async {
            self.eventSink?([
                "listenerId": self.listenerId,
                "event": "onAdShowed"
            ])
        }
    }
    
    func adDidClick(_ ad: PAGAdProtocol) {
        print("[PangleNativeAd] Ad clicked (\(listenerId))")
        DispatchQueue.main.async {
            self.eventSink?([
                "listenerId": self.listenerId,
                "event": "onAdClicked"
            ])
        }
    }
    
    func adDidDismiss(_ ad: PAGAdProtocol) {
        print("[PangleNativeAd] Ad dismissed (\(listenerId))")
    }
}
